viewer-prototype/src/browser/trace-server-status.ts

```diff
@@ -0,0 +1,101 @@
+/********************************************************************************
+ * Copyright (C) 2020 École Polytechnique de Montréal
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License v. 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0.
+ *
+ * This Source Code may also be made available under the following Secondary
+ * Licenses when the conditions for such availability set forth in the Eclipse
+ * Public License v. 2.0 are satisfied: GNU General Public License, version 2
+ * with the GNU Classpath Exception which is available at
+ * https://www.gnu.org/software/classpath/license.html.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+ ********************************************************************************/
+
+import { inject, injectable, postConstruct } from 'inversify';
+import { ILogger } from '@theia/core/lib/common/logger';
+import { DefaultFrontendApplicationContribution } from '@theia/core/lib/browser/frontend-application';
+import { StatusBar, StatusBarAlignment } from '@theia/core/lib/browser/status-bar/status-bar';
+import { Disposable, DisposableCollection } from '@theia/core/lib//common';
+import { ConnectionStatusService, ConnectionStatus, AbstractConnectionStatusService } from '@theia/core/lib/browser/connection-status-service';
+import { TspClient } from 'tsp-typescript-client/lib/protocol/tsp-client';
+
+@injectable()
+export class TraceServerConnectionStatusService extends AbstractConnectionStatusService {
+
+    private scheduledPing: number | undefined;
+
+    private constructor(
+        @inject(TspClient) private tspClient: TspClient
+    ) {
+        super();
+    }
+
+    @postConstruct()
+    protected init(): void {
+        this.schedulePing();
+    }
+
+    protected schedulePing(): void {
```

Comment:
Detecting lost connection did work for me when I stopped the server. But after restarting the server, then it didn't detect that the server is running again. I had to click into the "available analysis" area so that it got deblocked. Not sure what is the reason. I debugged a bit and added breakpoints in the methods below, but the breakpoint didn't hit until me clicking in the available analysis area.

Reply:
I got the same with the Theia server connection status (I tried all combinations of both server going offline), and only when I got back to the file explorer did the Theia server go back online... It's as if once offline, the ping doesn't get called again until the service is "required" by a view. I don't know where this behavior comes from. If you have exact steps to reproduce it always, please let me know, I'll try to figure it out in the meantime.

Reply:
It's already an improvement and gives feedback to the user.

Reply:
Leaving it as is for now, I'll investigate if/when I can reproduce with more consistency

Reply:
No wait, there's room for improvement here...